File: trnbias.jl
# using Revise
using Random
using Dates
using Random
using Printf

struct ParamsResult
    short_term::Int
    long_term::Int
    performance::Float64
end

mutable struct MarsagliaRng
    q::Vector{UInt32}
    carry::UInt32
    mwc256_initialized::Bool
    mwc256_seed::Int32
    i::UInt8

    function MarsagliaRng(seed::Vector{UInt8})
        q = zeros(UInt32, 256)
        carry = 362436
        mwc256_initialized = false
        mwc256_seed = reinterpret(Int32, seed)[1]
        i = 255
        new(q, carry, mwc256_initialized, mwc256_seed, i)
    end
end

# Default constructor using system time for seeding
function MarsagliaRng()
    ts_nano = UInt8(nanosecond(now()))
    seed = [ts_nano, ts_nano >>> 8, ts_nano >>> 16, ts_nano >>> 24]
    MarsagliaRng(seed)
end

# Random number generator functions
function next_u32(rng::MarsagliaRng)::UInt32
    a = UInt64(809430660)

    if !rng.mwc256_initialized
        j = UInt32(rng.mwc256_seed)
        c1 = UInt32(69069)
        c2 = UInt32(12345)
        rng.mwc256_initialized = true
        for k in 1:256
            j = (c1 * j + c2) % UInt32
            rng.q[k] = j
        end
    end

    rng.i = (rng.i + 1) % 256
    t = a * UInt64(rng.q[rng.i+1]) + UInt64(rng.carry)
    rng.carry = (t >> 32)
    rng.q[rng.i+1] = t % UInt32

    return rng.q[rng.i+1]
end



function next_u64(rng::MarsagliaRng)::UInt64
    UInt64(next_u32(rng))
end

# Sample function for generating a Float64
function rand(rng::MarsagliaRng)::Float64
    mult = 1.0 / typemax(UInt32)
    mult * next_u32(rng)
end



function opt_params(which::Int, ncases::Int, x::Vector{Float64})
    best_perf = -Inf
    ibestshort = 0
    ibestlong = 0

    for ilong in 2:199
        for ishort in 1:(ilong-1)
            total_return = 0.0
            win_sum = 1e-60
            lose_sum = 1e-60
            sum_squares = 1e-60
            short_sum = 0.0
            long_sum = 0.0

            for i in (ilong-1:ncases-2)
                if i == ilong-1
                    short_sum = sum(x[(i-ishort+2):i+1])
                    long_sum = sum(x[(i-ilong+2):i+1])
                else
                    short_sum += x[i+1] - x[i-ishort+1]
                    long_sum += x[i+1] - x[i-ilong+1]
                end


                short_mean = short_sum / ishort
                long_mean = long_sum / ilong

                ret = (short_mean > long_mean) ? x[i+2] - x[i+1] :
                      (short_mean < long_mean) ? x[i+1] - x[i+2] : 0.0

                total_return += ret
                sum_squares += ret^2
                if ret > 0.0
                    win_sum += ret
                else
                    lose_sum -= ret
                end
            end

            if which == 0
                total_return /= (ncases - ilong)
                if total_return > best_perf
                    best_perf = total_return
                    ibestshort = ishort
                    ibestlong = ilong
                end
            elseif which == 1
                pf = win_sum / lose_sum
                if pf > best_perf
                    best_perf = pf
                    ibestshort = ishort
                    ibestlong = ilong
                end
            elseif which == 2
                total_return /= (ncases - ilong)
                sum_squares /= (ncases - ilong)
                sum_squares -= total_return^2
                sr = total_return / (sqrt(sum_squares) + 1e-8)
                if sr > best_perf
                    best_perf = sr
                    ibestshort = ishort
                    ibestlong = ilong
                end
            end
        end
    end

    ParamsResult(ibestshort, ibestlong, best_perf)
end

function test_system(ncases::Int, x::Vector{Float64}, short_term::Int, long_term::Int)
    sum1 = 0.0

    for i in (long_term-1:ncases-2)
        short_mean = sum(x[i-short_term+2:i+1]) / short_term
        long_mean = sum(x[i-long_term+2:i+1]) / long_term

        if short_mean > long_mean
            sum1 += x[i+2] - x[i+1]
        elseif short_mean < long_mean
            sum1 -= x[i+2] - x[i+1]
        end

    end
    sum1 / (ncases - long_term)
end

function main()
    if length(ARGS) < 4
        println("Usage: julia script.jl <which> <ncases> <trend> <nreps>")
        return
    end

    println("sikeinanan")

    which = parse(Int, ARGS[1])
    ncases = parse(Int, ARGS[2])
    save_trend = parse(Float64, ARGS[3])
    nreps = parse(Int, ARGS[4])

    optimize(which, ncases, save_trend, nreps)
end

function optimize(which::Int, ncases::Int, save_trend::Float64, nreps::Int)
    x = zeros(Float64, ncases)

    is_mean = 0.0
    oos_mean = 0.0

    rng = MarsagliaRng(UInt8.([33, 0, 0, 0]))

    for irep in 1:nreps
        # Generate in-sample
        trend = save_trend
        x[1] = 0.0
        for i in 2:ncases
            if (i - 1) % 50 == 0
                trend = -trend
            end
            x[i] = x[i-1] + trend + rand(rng) + rand(rng) - rand(rng) - rand(rng)
        end

        params_result = opt_params(which, ncases, x)

        # Generate out-of-sample
        trend = save_trend
        x[1] = 0.0
        for i in 2:ncases
            if (i - 1) % 50 == 0
                trend = -trend
            end
            x[i] = x[i-1] + trend + rand(rng) + rand(rng) - rand(rng) - rand(rng)
        end


        oos_perf = test_system(ncases, x, params_result.short_term, params_result.long_term)

        is_mean += params_result.performance
        oos_mean += oos_perf
        # @printf("%3d: %3d %3d  %8.4f %8.4f (%8.4f)\n",
        #     irep, params_result.short_term, params_result.long_term, params_result.performance, oos_perf, params_result.performance - oos_perf)

    end

    is_mean /= nreps
    oos_mean /= nreps

    println("Mean IS=$is_mean  OOS=$oos_mean  Bias=$(is_mean - oos_mean)")

end

File: cpp_trnbias/trnbias.cc
/******************************************************************************/
/*                                                                            */
/*  TrnBias - Explore training bias                                           */
/*                                                                            */
/******************************************************************************/
 
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <float.h>
#include <stdlib.h>

// GJL added this line for gnu c++ on linux
#include <cstdint>

/*
--------------------------------------------------------------------------------
 
   This is a random int generator suggested by Marsaglia in his DIEHARD suite.
   It provides a great combination of speed and quality.
 
   We also have unifrand(), a random 0-1 generator based on it.
 
--------------------------------------------------------------------------------
*/
 
static unsigned int Q[256], carry=362436 ;
static int MWC256_initialized = 0 ;
static int MWC256_seed = 123456789 ;
static unsigned char i=255 ;
 
 
void RAND32M_seed ( int iseed ) { // Optionally set seed
   MWC256_seed = iseed ;
   MWC256_initialized = 0 ;
   carry = 362436 ;
   for (int k=0 ; k<256 ; k++) {
      Q[k] = 0;
   }
   i = 255;
   }
 
unsigned int RAND32M ()
{
   uint64_t t;
   uint64_t a=809430660 ;
 
   if (! MWC256_initialized) {
      unsigned int k,j=MWC256_seed ;
      MWC256_initialized = 1 ;
      for (k=0 ; k<256 ; k++) {
         j = 69069 * j + 12345 ; // This overflows, doing an automatic mod 2^32
         Q[k] = j ;
         }
      }
 
   i += 1;
   t = a * Q[i] + carry ;  // This is the 64-bit op, forced by a being 64-bit
   carry = (unsigned int) (t >> 32) ;
   Q[i] = (unsigned int) (t & 0xFFFFFFFF) ;
   return Q[i] ;
}
 
 
double unifrand ()
{
   static double mult = 1.0 / 0xFFFFFFFF ;
   return mult * RAND32M() ;
}
 
 
 
/*
--------------------------------------------------------------------------------
 
   Local routine computes optimal short-term and long-term lookbacks
   for a primitive moving-average crossover system
 
--------------------------------------------------------------------------------
*/
 
double opt_params (
   int which ,       // 0=mean return; 1=profit factor; 2=Sharpe ratio
   int ncases ,      // Number of log prices in X
   double *x ,       // Log prices
   int *short_term , // Returns optimal short-term lookback
   int *long_term    // Returns optimal long-term lookback
   )
{
   int i, j, ishort, ilong, ibestshort, ibestlong ;
   double short_sum, long_sum, short_mean, long_mean, total_return, best_perf ;
   double ret, win_sum, lose_sum, sum_squares, sr ;
 
   best_perf = -1.e60 ;                          // Will be best performance across all trials
   for (ilong=2 ; ilong<200 ; ilong++) {         // Trial long-term lookback
      for (ishort=1 ; ishort<ilong ; ishort++) { // Trial short-term lookback
 
         // We have a pair of lookbacks to try.
         // Cumulate performance for all valid cases
 
         total_return = 0.0 ;                    // Cumulate total return for this trial
         win_sum = lose_sum = 1.e-60 ;           // Cumulates for profit factor
         sum_squares = 1.e-60 ;                  // Cumulates for Sharpe ratio

         // GJL added these
         long_sum = 0.0;
         short_sum = 0.0;
         ibestlong = 0;
         ibestshort = 0;
 
         for (i=ilong-1 ; i<ncases-1 ; i++) {    // Compute performance across history
 
            if (i == ilong-1) { // Find the short-term and long-term moving averages for the first valid case.
               short_sum = 0.0 ;                 // Cumulates short-term lookback sum
               for (j=i ; j>i-ishort ; j--)
                  short_sum += x[j] ;
               long_sum = short_sum ;            // Cumulates long-term lookback sum
               while (j>i-ilong)
                  long_sum += x[j--] ;
               }
 
            else {                               // Update the moving averages
               short_sum += x[i] - x[i-ishort] ;
               long_sum += x[i] - x[i-ilong] ;
               }
 
            short_mean = short_sum / ishort ;
            long_mean = long_sum / ilong ;
 
            // We now have the short-term and long-term moving averages ending at day i
            // Take our position and cumulate performance
 
            if (short_mean > long_mean)       // Long position
               ret = x[i+1] - x[i] ;
            else if (short_mean < long_mean)  // Short position
               ret = x[i] - x[i+1] ;
            else
               ret = 0.0 ;
 
            total_return += ret ;
            sum_squares += ret * ret ;
            if (ret > 0.0)
               win_sum += ret ;
            else
               lose_sum -= ret ;
 
            } // For i, summing performance for this trial
 
         // We now have the performance figures across the history
         // Keep track of the best lookbacks
 
         if (which == 0) {      // Mean return criterion
            total_return /= ncases - ilong ;
            if (total_return > best_perf) {
               best_perf = total_return ;
               ibestshort = ishort ;
               ibestlong = ilong ;
               }
            }
 
         else if (which == 1  &&  win_sum / lose_sum > best_perf) { // Profit factor criterion
            best_perf = win_sum / lose_sum ;
            ibestshort = ishort ;
            ibestlong = ilong ;
            }
 
         else if (which == 2) { // Sharpe ratio criterion
            total_return /= ncases - ilong ;   // Now mean return
            sum_squares /= ncases - ilong ;
            sum_squares -= total_return * total_return ;  // Variance (may be zero!)
            sr = total_return / (sqrt ( sum_squares ) + 1.e-8) ;
            if (sr > best_perf) { // Profit factor criterion
               best_perf = sr ;
               ibestshort = ishort ;
               ibestlong = ilong ;
               }
            }
 
         } // For ishort, all short-term lookbacks
      } // For ilong, all long-term lookbacks
 
   *short_term = ibestshort ;
   *long_term = ibestlong ;
 
   return best_perf ;
}
 
 
/*
--------------------------------------------------------------------------------
 
   Local routine tests a trained crossover system
   This computes the mean return.  Users may wish to change it to
   compute other criteria.
 
--------------------------------------------------------------------------------
*/
 
double test_system (
   int ncases ,
   double *x ,
   int short_term ,
   int long_term
   )
{
   int i, j ;
   double sum, short_mean, long_mean ;
 
   sum = 0.0 ;                          // Cumulate performance for this trial
   for (i=long_term-1 ; i<ncases-1 ; i++) { // Sum performance across history
      short_mean = 0.0 ;                // Cumulates short-term lookback sum
      for (j=i ; j>i-short_term ; j--)
         short_mean += x[j] ;
      long_mean = short_mean ;          // Cumulates short-term lookback sum
      while (j>i-long_term)
         long_mean += x[j--] ;
      short_mean /= short_term ;
      long_mean /= long_term ;
      // We now have the short-term and long-term means ending at day i
      // Take our position and cumulate return
      if (short_mean > long_mean)       // Long position
         sum += x[i+1] - x[i] ;
      else if (short_mean < long_mean)  // Short position
         sum -= x[i+1] - x[i] ;
      } // For i, summing performance for this trial
   sum /= ncases - long_term ;              // Mean return across the history we just tested
 
   return sum ;
}
 
 
/*
--------------------------------------------------------------------------------
 
   Main routine
 
--------------------------------------------------------------------------------
*/
 
int main (
   int argc ,    // Number of command line arguments (includes prog name)
   char *argv[]  // Arguments (prog name is argv[0])
   )
{
   int i, which, ncases, irep, nreps, short_lookback, long_lookback ;
   double save_trend, trend, *x, IS_perf, OOS_perf, IS_mean, OOS_mean ;
 
/*
   Process command line parameters
*/
 
#if 1
   if (argc != 5) {
      printf ( "\nUsage: TrnBias  which  ncases trend  nreps" ) ;
      printf ( "\n  which - 0=mean return  1=profit factor  2=Sharpe ratio" ) ;
      printf ( "\n  ncases - number of training and test cases" ) ;
      printf ( "\n  trend - Amount of trending, 0 for flat system" ) ;
      printf ( "\n  nreps - number of test replications" ) ;
      exit ( 1 ) ;
      }
 
   which = atoi ( argv[1] ) ;
   ncases = atoi ( argv[2] ) ;
   save_trend = atof ( argv[3] ) ;
   nreps = atoi ( argv[4] ) ;
#else
   which = 1 ;
   ncases = 1000 ;
   save_trend = 0.2 ;
   nreps = 10 ;
#endif
 
   if (ncases < 2  ||  which < 0  ||  which > 2  ||  nreps < 1) {
      printf ( "\nUsage: TrnBias  which  ncases trend  nreps" ) ;
      printf ( "\n  which - 0=mean return  1=profit factor  2=Sharpe ratio" ) ;
      printf ( "\n  ncases - number of training and test cases" ) ;
      printf ( "\n  trend - Amount of trending, 0 for flat system" ) ;
      printf ( "\n  nreps - number of test replications" ) ;
      exit ( 1 ) ;
      }
 
   printf ( "\n\nwhich=%d ncases=%d trend=%.3lf nreps=%d", which, ncases, save_trend, nreps ) ;
 
/*
   Initialize
*/
 
   x = (double *) malloc ( ncases * sizeof(double) ) ;
 
/*
   Main replication loop
*/
 
   IS_mean = OOS_mean = 0.0 ;
   for (irep=0 ; irep<nreps ; irep++) {  // Do many trials to get a stable average
 
      // Generate the in-sample set (log prices)
      trend = save_trend ;
      x[0] = 0.0 ;
      for (i=1 ; i<ncases ; i++) {
         if (i % 50 == 0) // Reverse the trend
            trend = -trend ;
         x[i] = x[i-1] + trend + unifrand() + unifrand() - unifrand() - unifrand() ;
         }
 
      // Compute optimal parameters, evaluate return with same dataset
      opt_params ( which , ncases , x , &short_lookback , &long_lookback ) ;
      IS_perf = test_system ( ncases , x , short_lookback , long_lookback ) ;
 
      // Generate the out_of-sample set (log prices)
      trend = save_trend ;
      x[0] = 0.0 ;
      for (i=1 ; i<ncases ; i++) {
         if (i % 50 == 0)
            trend = -trend ;
         x[i] = x[i-1] + trend + unifrand() + unifrand() - unifrand() - unifrand() ;
         }
 
      // Test the OOS set and cumulate means across replications
      OOS_perf = test_system ( ncases , x , short_lookback , long_lookback ) ;
 
      IS_mean += IS_perf ;
      OOS_mean += OOS_perf ;
      //      printf ( "\n%3d: %3d %3d  %8.4lf %8.4lf (%8.4lf)", irep, short_lookback, long_lookback, IS_perf, OOS_perf, IS_perf - OOS_perf ) ;
      } // For irep
 
   // Done.  Print results and clean up.
   IS_mean /= nreps ;
   OOS_mean /= nreps ;
   printf ( "\nMean IS=%.16lf  OOS=%.16lf  Bias=%.16lf\n", IS_mean, OOS_mean, IS_mean - OOS_mean ) ;
   //   printf ( "\nMean IS=%.4lf  OOS=%.4lf  Bias=%.4lf", IS_mean, OOS_mean, IS_mean - OOS_mean ) ;
 
   free ( x ) ;
 
   return 0 ;
}
